docs/design/apiserver_storage.md

```diff
@@ -0,0 +1,216 @@
+
+	// WatchPrefix watches a prefix after given version. If version is 0, we will watch from current state.
+	// It returns notifications of any keys that has given prefix.
+	// Given selectors, it returns events that contained object of interest, either of current and previous.
+	// If there is any problem establishing the watch channel, it will return error.
+	// After channel is established, any error will be returned from WatchChan immediately before it's closed.
+	WatchPrefix(ctx context.Context, prefix StorageKey, version int64, ss ...Selector) (WatchChan, error)
+
+	// AddIndex adds a new index.
+	// An index indicates a field(s) of a type of an object that would be queried or watched frequently and
+	// thus require better efficiency and performance.
+	// If an index exists for a field, complexity of a query on it should be no more than O(logN)
+	// where N is the total number of objects for List or watchers for watch event delivery.
+	// See FieldValueGetFunc for more details on field value retrieving.
+	AddIndex(indexName, field string, g FieldValueGetFunc) error
+	// DeleteIndex deletes an index.
+	DeleteIndex(indexName string)
+}
+
+// KeyResponse contains the object as a response for a key.
+type KeyResponse struct {
+	Object runtime.Object
+	// Version is the storage version of the object in this response.
+	Version int64
+}
+
+// ListResponse contains objects as a response for a prefix.
+type ListResponse struct {
+	Keys []*KeyResponse
+	// HeadVersion is the latest version known by this storage.
+	HeadVersion int64
+}
+
+// Conditions is used in do atomic conditional operations.
+// If compare failed, corresponding method should return storage version conflict error.
+type Conditions struct {
+	// PrevVersion is used to compare with existing object’s version.
+	// Compare would only succeed if versions were matched.
+	// Note that PrevVersion=0 means no previous object existed.
+	PrevVersion int64
+}
+
+// FieldValueGetFunc returns the value in the object corresponding to given field.
+// field is uniquely indicating a field of a type of object.
+// It returns (value, true) if field exists in the object. Otherwise it returns ("", false).
+type FieldValueGetFunc func(field string, obj runtime.Object) (string, bool)
+
+// Selector is used to express one selection predicate.
+// Op is the operator relating corresponding field's value to given values.
+// See FieldValueGetFunc for details on Field and FVGetFunc.
+type Selector struct {
+	Op        Operator
+	Field     string
+	Values    []string
+	FVGetFunc FieldValueGetFunc
+}
+
+type WatchChan <-chan WatchResponse
+
+type WatchResponse struct {
+	Type       EventType
+	Object     runtime.Object
+	PrevObject runtime.Object
+	Err        error
+}
+
+// StorageKey consists of multiple paths and after being joined by "/" become a unique path.
+type StorageKey []string
+```
+
+We made the following changes based on the current Storage.Interface:
+
+#### 1. More index friendly
+
```

Comment:
The index part is which I am mainly interested.  
So, now we can pass a filter function of the Object value to the etcd in list and watch?
what's the interface looks like? as I know now etcd only support key index rather than value index